tests/integration/security/authn/jwt/jwt_test.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jwt
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/test/echo/common/scheme"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/echo/echoboot"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+	"istio.io/istio/pkg/test/util/file"
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/istio/pkg/test/util/tmpl"
+	"istio.io/istio/tests/integration/security/util"
+	"istio.io/istio/tests/integration/security/util/connection"
+)
+
+const (
+	authHeaderKey = "Authorization"
+	validJwt      = "eyJhbGciOiJSUzI1NiIsImtpZCI6IkRIRmJwb0lVcXJZOHQyenBBMnFYZkNtcjVWTzVaRX" +
+		"I0UnpIVV8tZW52dlEiLCJ0eXAiOiJKV1QifQ.eyJleHAiOjM3NTU5ODU5MzQ2LCJpYXQiOjE1NTk4NTkzNDYsImlzcyI" +
+		"6InRlc3RpbmdAc2VjdXJlLmlzdGlvLmlvIiwic3ViIjoidGVzdGluZ0BzZWN1cmUuaXN0aW8uaW8ifQ.A4j9ft49KrKw" +
+		"zvgpzoKXdMxVyYTUjuU3LMgqjvfZXDSUIYtL6gx31CSMb2UZiESmJ7Xu8XkqhcZvmfcuU_WApl01emBOwjwtg50M-Yuc" +
+		"oy6A4DbnsVcKnSY7VtAWrJ9ACdr1CUPseTfkbaIW-PYDfLPTMJljO2NdzF0yC9l97N8lau9afNu9Ilc9DIlQ5CgW6h1J" +
+		"VasV21wSLlrglnWhut6yV6yQOqUjMcwgntgVkv_3UbdtzJKt46u29Juh_8m4OEdAXeyy00midPSIJWIm9J6lgLF35XUB" +
+		"e_nvonKwdsxKpV-LrWlKMZB8QneHbb3194J7Wr2H7WVnNGcbVUCORw"
+	expiredJwt = "eyJhbGciOiJSUzI1NiIsImtpZCI6IkRIRmJwb0lVcXJZOHQyenBBMnFYZkNtcjVWTzVaRXI0UnpIVV8tZ" +
+		"W52dlEiLCJ0eXAiOiJKV1QifQ.eyJleHAiOjE1NTk5MzEyMDksImlhdCI6MTU1OTkzMTIwOSwiaXNzIjoidGVzdGluZ0" +
+		"BzZWN1cmUuaXN0aW8uaW8iLCJzdWIiOiJ0ZXN0aW5nQHNlY3VyZS5pc3Rpby5pbyJ9.a7zuyoF7eaVcnHOkJKK25WqTd" +
+		"K4OT0sgQzrAqYhANv-6MfbIHgRtbOYvQW0pSL5saeSkpd8YV35NdmipsbPcMTVYgSlSPkmNdinZwJyGpqdGEu6fVYq3P" +
+		"FET0bBSm5yVTkO7yyX8AgVVH31ouGJ8OQ11gGZ66Jmle5PNwyGCh1ccZsT8LefYTbcDHMbXnoYwU4e3WwcphLBqFoafF" +
+		"JUpRXV5dtn2YLnMwA0ALTteRMVMYIkXQkR6QhBwUufC3aUQmZydzrGMaKqbwcYbcp1GG05v4A99rikNQ-Ia6xswgAEJu" +
+		"JaYhppL-0B7E-i4jhpGbKOrFyc6vFZlk9oejUpi7Q"
+)
+
+func TestAuthnJwt(t *testing.T) {
+	framework.NewTest(t).
+		Run(func(ctx framework.TestContext) {
+			ns := namespace.NewOrFail(t, ctx, "authn-jwt", true)
+
+			var a, b, c echo.Instance
+			echoboot.NewBuilderOrFail(ctx, ctx).
+				With(&a, util.EchoConfig("a", ns, false, nil, g, p)).
+				With(&b, util.EchoConfig("b", ns, false, nil, g, p)).
+				With(&c, util.EchoConfig("c", ns, false, nil, g, p)).
+				BuildOrFail(t)
+
+			testCases := []struct {
```

Comment:
Can we just first migrate the existing jwt e2e tests to the new framework? https://github.com/istio/istio/blob/83102c738251e2bfdb8cf18002d1ca4c674a0aa5/tests/e2e/tests/pilot/authn_policy_test.go#L101

This helps to make sure we have the same coverage in the 2 test framework, and we can add more tests (only in the new framework) after.

Reply:
Tests from the old framework are also run besides in the new framework. Also, the jwt test here https://github.com/istio/istio/blob/49c978439ddde62f4d5e0cf672976294797cf2fa/tests/e2e/tests/pilot/testdata/authn/v1alpha1/authn-policy-jwt.yaml.tmpl is pretty basic. I think this new one cover more cases, but I can look into the old test as a follow up PR if you think we need to

Reply:
The point is to first have same coverage then increase more cases, so when we later remove the old e2e test, we don't worry about losing any coverage, and the easier way to do this is starting from the same test in the new test framework.

One thing not covered in the new test framework, is the case for https://github.com/istio/istio/blob/83102c738251e2bfdb8cf18002d1ca4c674a0aa5/tests/e2e/tests/pilot/authn_policy_test.go#L153, note we specify two jwt issuers in the same jwt policy and make sure only it's used in different paths at the same time.

Reply:
Is that the only difference (besides the ingress one)? If so, I can ask that test case to this PR.

Reply:
I'm not sure, you need to look into the code, that's what I mean it's harder to figure out if we start by adding new tests, anyway, you can choose either of the following:
- Add new tests and then migrate the old tests
- Migrate old tests and then add new tests
Either way should make sure when we later remove the old test code, we will have the same coverage in the new test.